Flatten step potentially creates corrupted avro records

To reproduce: (in docker env) 

docker-compose.yaml
```
postgres:
    image: postgres:14.3
    container_name: postgres
    networks: [ "net1" ]
    environment:
      - POSTGRES_PASSWORD=password
      - POSTGRES_USER=postgres
    ports:
      - 5432:5432
    command:
      - "postgres"
      - "-c"
      - "wal_level=logical"
```

1. Setup postgre debezium source with avro converter
```
docker-compose up -d postgres

docker exec -it postgres psql -U postgres -c 'create table if not exists pulsar_table_source (key VARCHAR(255) PRIMARY KEY, value VARCHAR(255) NULL);' postgres



docker exec -it pulsar bin/pulsar-admin sources create --namespace default --tenant public \
    --source-type  debezium-postgres \
    --destination-topic-name debezium-postgres-topic \
    --name pg-source \
    --source-config '{"key.converter": "org.apache.pulsar.kafka.shade.io.confluent.connect.avro.AvroConverter", "value.converter": "org.apache.pulsar.kafka.shade.io.confluent.connect.avro.AvroConverter", "plugin.name": "pgoutput","database.hostname": "postgres", "database.user": "postgres", "database.server.name": "dbserver1", "database.password": "password", "database.dbname": "postgres", "schema.whitelist": "public", "pulsar.service.url": "pulsar://pulsar:6650" }' \
    --parallelism 1

docker exec -it postgres bash -c "psql -U postgres -c \"insert into pulsar_source_table(key,value) select left(md5(i::text), 10),i from generate_series(1, 10) s(i)\""

```

2. Create a sink (in my case a Pulsar Sink) with the transforms fn attached
```
docker exec -it pulsar bin/pulsar-admin sinks create --namespace default --tenant public \
    --sink-type cassandra-enhanced \
    --name cassandra-sink \
    --inputs dbserver1.public.pulsar_source_table \
    --subs-position Earliest \
    --sink-config '{"verbose": true, "tasks.max":1,"topics":"dbserver1.public.pulsar_source_table","contactPoints":"cassandra","loadBalancing.localDc":"datacenter1","port":9042,"cloud.secureConnectBundle":null,"ignoreErrors":"None","maxConcurrentRequests":500,"maxNumberOfRecordsInBatch":32,"queryExecutionTimeout":30,"connectionPoolLocalSize":4,"jmx":true,"compression":"None","auth":{"provider":"None","username":null,"password":null,"gssapi":{"keyTab":null,"principal":null,"service":"dse"}},"ssl":{"provider":"None","hostnameValidation":true,"keystore":{"password":null,"path":null},"openssl":{"keyCertChain":null,"privateKey":null},"truststore":{"password":null,"path":null},"cipherSuites":null},"topic":{"dbserver1.public.pulsar_source_table":{"baselines":{"pgsourcetable":{"mapping":"key=key.key, value=value.myvalue","consistencyLevel":"LOCAL_ONE","ttl":-1,"ttlTimeUnit":"SECONDS","timestampTimeUnit":"MICROSECONDS","nullToUnset":true,"deletesEnabled":true}},"codec":{"locale":"en_US","timeZone":"UTC","timestamp":"CQL_TIMESTAMP","date":"ISO_LOCAL_DATE","time":"ISO_LOCAL_TIME","unit":"MILLISECONDS"}}}}' \
    --parallelism 1 \
    --transform-function builtin://transform \
    --transform-function-config '{"steps": [{"type": "flatten", "part": "value"}]}'
```
Then you get this error in the sink
```
org.apache.pulsar.client.api.SchemaSerializationException: java.io.EOFException
        at org.apache.pulsar.client.impl.schema.generic.GenericAvroReader.read(GenericAvroReader.java:91) ~[?:?]
        at org.apache.pulsar.client.impl.schema.generic.GenericAvroReader.read(GenericAvroReader.java:41) ~[?:?]
        at org.apache.pulsar.client.api.schema.SchemaReader.read(SchemaReader.java:40) ~[java-instance.jar:?]
        at org.apache.pulsar.client.impl.schema.reader.AbstractMultiVersionReader.read(AbstractMultiVersionReader.java:89) ~[?:?]
        at org.apache.pulsar.client.impl.schema.AbstractStructSchema.decode(AbstractStructSchema.java:66) ~[?:?]
        at org.apache.pulsar.client.impl.schema.KeyValueSchemaImpl.decode(KeyValueSchemaImpl.java:180) ~[?:?]
        at org.apache.pulsar.client.impl.schema.KeyValueSchemaImpl.lambda$decode$0(KeyValueSchemaImpl.java:161) ~[?:?]
        at org.apache.pulsar.common.schema.KeyValue.decode(KeyValue.java:146) ~[java-instance.jar:?]
        at org.apache.pulsar.client.impl.schema.KeyValueSchemaImpl.decode(KeyValueSchemaImpl.java:161) ~[?:?]
        at org.apache.pulsar.client.impl.schema.KeyValueSchemaImpl.decode(KeyValueSchemaImpl.java:43) ~[?:?]
        at org.apache.pulsar.client.impl.schema.AutoConsumeSchema.decode(AutoConsumeSchema.java:120) ~[?:?]
        at org.apache.pulsar.client.impl.schema.AutoConsumeSchema.decode(AutoConsumeSchema.java:48) ~[?:?]
        at org.apache.pulsar.functions.instance.JavaInstanceRunnable.encodeWithRecordSchemaAndDecodeWithSinkSchema(JavaInstanceRunnable.java:459) ~[?:?]
        at org.apache.pulsar.functions.instance.JavaInstanceRunnable.sendOutputMessage(JavaInstanceRunnable.java:416) ~[?:?]
        at org.apache.pulsar.functions.instance.JavaInstanceRunnable.handleResult(JavaInstanceRunnable.java:394) ~[?:?]
        at org.apache.pulsar.functions.instance.JavaInstanceRunnable.run(JavaInstanceRunnable.java:335) ~[?:?]
        at java.lang.Thread.run(Thread.java:829) ~[?:?]
Caused by: java.io.EOFException
        at [org.apache.avro.io](http://org.apache.avro.io/).BinaryDecoder$ByteArrayByteSource.readRaw(BinaryDecoder.java:999) ~[java-instance.jar:?]
        at [org.apache.avro.io](http://org.apache.avro.io/).BinaryDecoder.doReadBytes(BinaryDecoder.java:405) ~[java-instance.jar:?]
        at [org.apache.avro.io](http://org.apache.avro.io/).BinaryDecoder.readString(BinaryDecoder.java:313) ~[java-instance.jar:?]
        at [org.apache.avro.io](http://org.apache.avro.io/).ResolvingDecoder.readString(ResolvingDecoder.java:208) ~[java-instance.jar:?]
        at org.apache.avro.generic.GenericDatumReader.readString(GenericDatumReader.java:469) ~[java-instance.jar:?]
        at org.apache.avro.generic.GenericDatumReader.readString(GenericDatumReader.java:459) ~[java-instance.jar:?]
        at org.apache.avro.generic.GenericDatumReader.readWithoutConversion(GenericDatumReader.java:191) ~[java-instance.jar:?]
        at org.apache.avro.generic.GenericDatumReader.read(GenericDatumReader.java:160) ~[java-instance.jar:?]
        at org.apache.avro.generic.GenericDatumReader.readField(GenericDatumReader.java:259) ~[java-instance.jar:?]
        at org.apache.avro.generic.GenericDatumReader.readRecord(GenericDatumReader.java:247) ~[java-instance.jar:?]
        at org.apache.avro.generic.GenericDatumReader.readWithoutConversion(GenericDatumReader.java:179) ~[java-instance.jar:?]
        at org.apache.avro.generic.GenericDatumReader.read(GenericDatumReader.java:160) ~[java-instance.jar:?]
        at org.apache.avro.generic.GenericDatumReader.read(GenericDatumReader.java:153) ~[java-instance.jar:?]
        at org.apache.pulsar.client.impl.schema.generic.GenericAvroReader.read(GenericAvroReader.java:86) ~[?:?]
        ... 16 more
```

Notes:
1. If I remove the transform function, the sink is able to read the record coming from the debezium source
2. I use the Avro format because JSON is not supported by the flatten step (perhaps we should track an issue)
3. My first assumption is that the output schema is not keeping the schema props. the debezium source set a special prop "__AVRO_READ_OFFSET__" to 5. This is used in Pulsar to set the start offset for reading the record. I don't know why and it may cause such error btw

--- a/pulsar-transformations/src/main/java/com/datastax/oss/pulsar/functions/transforms/FlattenStep.java
+++ b/pulsar-transformations/src/main/java/com/datastax/oss/pulsar/functions/transforms/FlattenStep.java
@@ -170,17 +170,16 @@ public class FlattenStep implements TransformStep {
       return null;
     }
 
-    if (schema.getField(name).schema().isUnion()) {
-      for (org.apache.avro.Schema s : schema.getField(name).schema().getTypes()) {
+    org.apache.avro.Schema fieldSchema = schema.getField(name).schema();
+    if (fieldSchema.isUnion()) {
+      for (org.apache.avro.Schema s : fieldSchema.getTypes()) {
         if (s.getType() == org.apache.avro.Schema.Type.RECORD) {
           return s;
         }
       }
     }
 
-    return schema.getField(name).schema().getType() == org.apache.avro.Schema.Type.RECORD
-        ? schema.getField(name).schema()
-        : null;
+    return fieldSchema.getType() == org.apache.avro.Schema.Type.RECORD ? fieldSchema : null;
   }
 
   /**

--- a/pulsar-transformations/src/main/java/com/datastax/oss/pulsar/functions/transforms/FlattenStep.java
+++ b/pulsar-transformations/src/main/java/com/datastax/oss/pulsar/functions/transforms/FlattenStep.java
@@ -18,8 +18,10 @@ package com.datastax.oss.pulsar.functions.transforms;
 import java.util.ArrayList;
 import java.util.List;
 import java.util.stream.Collectors;
+import javax.annotation.Nullable;
 import lombok.Builder;
 import lombok.Value;
+import org.apache.avro.SchemaBuilder;
 import org.apache.avro.generic.GenericData;
 import org.apache.avro.generic.GenericRecord;
 import org.apache.pulsar.client.api.Schema;
@@ -31,6 +33,8 @@ public class FlattenStep implements TransformStep {
   // TODO: Microbenchmark the flatten algorithm for performance optimization
   // TODO: Validate flatten delimiter
   // TODO: Add integration test
+
+  public static final String AVRO_READ_OFFSET_PROP = "__AVRO_READ_OFFSET__";
 
   private static final String DEFAULT_DELIMITER = "_"; // '.' in not valid in AVRO field names
 
@@ -88,7 +92,8 @@ public class FlattenStep implements TransformStep {
     List<FieldValuePair> flattenedFields = new ArrayList<>();
     org.apache.avro.Schema originalSchema = record.getSchema();
     for (org.apache.avro.Schema.Field field : originalSchema.getFields()) {
-      flattenedFields.addAll(flattenField(record, field, ""));
+      flattenedFields.addAll(
+          flattenField(record, record.getSchema(), field, field.schema().isNullable(), ""));
     }
 
     return flattenedFields;
@@ -97,49 +102,102 @@ public class FlattenStep implements TransformStep {
   org.apache.avro.Schema buildFlattenedSchema(
       GenericRecord record, List<org.apache.avro.Schema.Field> flattenedFields) {
     org.apache.avro.Schema originalSchema = record.getSchema();
-
-    return org.apache.avro.Schema.createRecord(
-        originalSchema.getName(),
-        originalSchema.getDoc(),
-        originalSchema.getNamespace(),
-        false,
-        flattenedFields);
+    org.apache.avro.Schema flattenedSchema =
+        org.apache.avro.Schema.createRecord(
+            originalSchema.getName(),
+            originalSchema.getDoc(),
+            originalSchema.getNamespace(),
+            false,
+            flattenedFields);
+    originalSchema
+        .getObjectProps()
+        .entrySet()
+        .stream()
+        .filter(e -> !AVRO_READ_OFFSET_PROP.equals(e.getKey()))
+        .forEach(e -> flattenedSchema.addProp(e.getKey(), e.getValue()));
+    return flattenedSchema;
   }
 
+  /**
+   * @param record the record that contains the field to be flattened. Each recursive call will pass
+   *     the record one level deeper. At any level, the field could become null.
+   * @param schema the schema of the record that contains the field to be flattened. Each recursive
+   *     call will pass the schema one level deeper. At any level, the schema should exist.
+   * @param field the field to be flattened
+   * @param nullable true if the current field schema is nullable, this has to be passed all the way
+   *     to the last nested level because anytime one of the ancestors is null, the flattened field
+   *     schema could be null even if it is not nullable on the original schema
+   * @param flattenedFieldName the field name that is built incrementally with each recursive call.
+   * @return flattened key/value pairs.
+   */
   List<FieldValuePair> flattenField(
-      GenericRecord record, org.apache.avro.Schema.Field field, String flattenedFieldName) {
+      @Nullable GenericRecord record,
+      org.apache.avro.Schema schema,
+      org.apache.avro.Schema.Field field,
+      boolean nullable,
+      String flattenedFieldName) {
     List<FieldValuePair> flattenedFields = new ArrayList<>();
     // Because of UNION schemas, we cannot tell for sure if the current field is a record, so we
     // have to use reflection
-    if (record.get(field.name()) instanceof GenericRecord) {
-      GenericRecord genericRecord = (GenericRecord) record.get(field.name());
-      for (org.apache.avro.Schema.Field nestedField : genericRecord.getSchema().getFields()) {
+    org.apache.avro.Schema recordSchema = getRecordSchema(schema, field.name());
+    if (recordSchema != null) {
+      for (org.apache.avro.Schema.Field nestedField : recordSchema.getFields()) {
+        GenericRecord genericRecord =
+            record == null ? null : (GenericRecord) record.get(field.name());
         flattenedFields.addAll(
             flattenField(
-                genericRecord, nestedField, flattenedFieldName + field.name() + delimiter));
+                genericRecord,
+                recordSchema,
+                nestedField,
+                nullable || nestedField.schema().isNullable(),
+                flattenedFieldName + field.name() + delimiter));
       }
 
       return flattenedFields;
     }
 
     org.apache.avro.Schema.Field flattenedField =
-        createField(field, flattenedFieldName + field.name());
-    FieldValuePair fieldValuePair = new FieldValuePair(flattenedField, record.get(field.name()));
+        createField(field, flattenedFieldName + field.name(), nullable);
+    FieldValuePair fieldValuePair =
+        new FieldValuePair(flattenedField, record == null ? null : record.get(field.name()));
     flattenedFields.add(fieldValuePair);
 
     return flattenedFields;
+  }
+
+  private org.apache.avro.Schema getRecordSchema(org.apache.avro.Schema schema, String name) {
+    if (schema.getType() != org.apache.avro.Schema.Type.RECORD) {
+      return null;
+    }
+
+    if (schema.getField(name).schema().isUnion()) {
+      for (org.apache.avro.Schema s : schema.getField(name).schema().getTypes()) {
+        if (s.getType() == org.apache.avro.Schema.Type.RECORD) {
+          return s;
+        }
+      }
+    }
+
+    return schema.getField(name).schema().getType() == org.apache.avro.Schema.Type.RECORD
+        ? schema.getField(name).schema()
+        : null;
   }
 
   /**
    * Creates a new Field instance with the same schema, doc, defaultValue, and order as field has
    * with changing the name to the specified one. It also copies all the props and aliases.
    */
-  org.apache.avro.Schema.Field createField(org.apache.avro.Schema.Field field, String name) {
+  org.apache.avro.Schema.Field createField(
+      org.apache.avro.Schema.Field field, String name, boolean nullable) {
+    org.apache.avro.Schema newSchema = field.schema();
+    if (nullable && !newSchema.isNullable()) {
+      newSchema = SchemaBuilder.unionOf().nullType().and().type(newSchema).endUnion();
+    }
     org.apache.avro.Schema.Field newField =
         new org.apache.avro.Schema.Field(
-            name, field.schema(), field.doc(), field.defaultVal(), field.order());
+            name, newSchema, field.doc(), field.defaultVal(), field.order());
     newField.putAll(field);
-    field.aliases().forEach(field::addAlias);
+    field.aliases().forEach(newField::addAlias);
 
     return newField;
   }

--- a/pulsar-transformations/src/test/java/com/datastax/oss/pulsar/functions/transforms/FlattenStepTest.java
+++ b/pulsar-transformations/src/test/java/com/datastax/oss/pulsar/functions/transforms/FlattenStepTest.java
@@ -15,10 +15,14 @@
  */
 package com.datastax.oss.pulsar.functions.transforms;
 
+import static com.datastax.oss.pulsar.functions.transforms.FlattenStep.AVRO_READ_OFFSET_PROP;
 import static org.testng.Assert.assertEquals;
+import static org.testng.Assert.assertNull;
 import static org.testng.Assert.assertSame;
 
+import java.util.Map;
 import java.util.Optional;
+import java.util.stream.Collectors;
 import org.apache.avro.Schema;
 import org.apache.avro.generic.GenericData;
 import org.apache.avro.util.Utf8;
@@ -123,7 +127,7 @@ public class FlattenStepTest {
   }
 
   @Test
-  void testNestedKeyValueFlattenedWithCustomerDelimiter() throws Exception {
+  void testNestedKeyValueFlattenedWithCustomDelimiter() throws Exception {
     // given
     Record<GenericObject> nestedKVRecord = Utils.createNestedAvroKeyValueRecord(4);
 
@@ -278,8 +282,8 @@ public class FlattenStepTest {
 
   private void assertSchemasFlattened(
       GenericData.Record actual, GenericData.Record expected, String delimiter) {
-    assertEquals(actual.getSchema().getFields().size(), 9);
-
+    assertEquals(actual.getSchema().getFields().size(), 11);
+    assertTopLevelSchema(actual.getSchema(), expected.getSchema());
     assertField(
         actual.getSchema().getField("level1String"), expected.getSchema().getField("level1String"));
     Schema.Field expectedL1 = expected.getSchema().getField("level1Record");
@@ -319,9 +323,9 @@ public class FlattenStepTest {
             .getSchema()
             .getField(
                 String.format(
-                    "level1Record%1$slevel2Record%1$slevel3Record%1$slevel4StringWithProps",
+                    "level1Record%1$slevel2Record%1$slevel3Record%1$slevel4StringWithPropsAndAlias",
                     delimiter)),
-        expectedL3.schema().getField("level4StringWithProps"));
+        expectedL3.schema().getField("level4StringWithPropsAndAlias"));
     assertField(
         actual
             .getSchema()
@@ -337,7 +341,7 @@ public class FlattenStepTest {
 
   private void assertValuesFlattened(
       GenericData.Record actual, GenericData.Record expected, String delimiter) {
-    assertEquals(actual.getSchema().getFields().size(), 9);
+    assertEquals(actual.getSchema().getFields().size(), 11);
 
     assertEquals(actual.get("level1String"), new Utf8((String) expected.get("level1String")));
     GenericData.Record expectedL1 = (GenericData.Record) expected.get("level1Record");
@@ -367,14 +371,23 @@ public class FlattenStepTest {
     assertEquals(
         actual.get(
             String.format(
-                "level1Record%1$slevel2Record%1$slevel3Record%1$slevel4StringWithProps",
+                "level1Record%1$slevel2Record%1$slevel3Record%1$slevel4StringWithPropsAndAlias",
                 delimiter)),
-        new Utf8((String) expectedL3.get("level4StringWithProps")));
+        new Utf8((String) expectedL3.get("level4StringWithPropsAndAlias")));
     assertEquals(
         actual.get(
             String.format(
                 "level1Record%1$slevel2Record%1$slevel3Record%1$slevel4Union", delimiter)),
         new Utf8((String) expectedL3.get("level4Union")));
+    assertNull(
+        actual.get(
+            String.format(
+                "level1Record%1$slevel2Record%1$slevel3Record%1$slevel4Null", delimiter)));
+    assertNull(
+        actual.get(
+            String.format(
+                "level1Record%1$slevel2Record%1$slevel3Record%1$slevel4NullRecord%1$signored",
+                delimiter)));
   }
 
   private void assertValuesFlattened(GenericData.Record actual, GenericData.Record expected) {
@@ -386,5 +399,20 @@ public class FlattenStepTest {
     assertEquals(actual.doc(), expected.doc());
     assertEquals(actual.defaultVal(), expected.defaultVal());
     assertEquals(actual.getObjectProps(), expected.getObjectProps());
+    assertEquals(actual.aliases(), expected.aliases());
+  }
+
+  private void assertTopLevelSchema(Schema actual, Schema expected) {
+    assertEquals(actual.getName(), expected.getName());
+    assertEquals(actual.getNamespace(), expected.getNamespace());
+    assertEquals(actual.getDoc(), expected.getDoc());
+    assertEquals(
+        actual.getObjectProps(),
+        expected
+            .getObjectProps()
+            .entrySet()
+            .stream()
+            .filter(e -> !AVRO_READ_OFFSET_PROP.equals(e.getKey()))
+            .collect(Collectors.toMap(Map.Entry::getKey, Map.Entry::getValue)));
   }
 }

--- a/pulsar-transformations/src/test/java/com/datastax/oss/pulsar/functions/transforms/Utils.java
+++ b/pulsar-transformations/src/test/java/com/datastax/oss/pulsar/functions/transforms/Utils.java
@@ -15,6 +15,7 @@
  */
 package com.datastax.oss.pulsar.functions.transforms;
 
+import static com.datastax.oss.pulsar.functions.transforms.FlattenStep.AVRO_READ_OFFSET_PROP;
 import static org.testng.Assert.assertEquals;
 import static org.testng.Assert.assertFalse;
 import static org.testng.Assert.assertTrue;
@@ -210,6 +211,17 @@ public class Utils {
     List<org.apache.avro.Schema.Field> fields = new ArrayList<>();
     org.apache.avro.Schema nullAndString =
         SchemaBuilder.unionOf().stringType().and().nullType().endUnion();
+    org.apache.avro.Schema nullLevel =
+        org.apache.avro.Schema.createRecord(
+            "nullLevel",
+            "doc ",
+            "ns",
+            false,
+            List.of(
+                new org.apache.avro.Schema.Field(
+                    "ignored", org.apache.avro.Schema.create(org.apache.avro.Schema.Type.STRING))));
+    org.apache.avro.Schema nullAndRecord =
+        SchemaBuilder.unionOf().nullType().and().type(nullLevel).endUnion();
     fields.add(
         new org.apache.avro.Schema.Field(
             "level" + levels + "String",
@@ -236,12 +248,16 @@ public class Utils {
             "level" + levels + "Array", arraySchema, "arrayDoc", Collections.emptyList()));
     org.apache.avro.Schema.Field stringWithProps =
         new org.apache.avro.Schema.Field(
-            "level" + levels + "StringWithProps",
+            "level" + levels + "StringWithPropsAndAlias",
             org.apache.avro.Schema.create(org.apache.avro.Schema.Type.STRING));
     stringWithProps.addProp("p1", "v1");
     stringWithProps.addProp("p2", "v2");
+    stringWithProps.addProp(AVRO_READ_OFFSET_PROP, "5");
+    stringWithProps.addAlias("a1");
     fields.add(stringWithProps);
     fields.add(new org.apache.avro.Schema.Field("level" + levels + "Union", nullAndString));
+    fields.add(new org.apache.avro.Schema.Field("level" + levels + "Null", nullAndString));
+    fields.add(new org.apache.avro.Schema.Field("level" + levels + "NullRecord", nullAndRecord));
     org.apache.avro.Schema lastLevelSchema =
         org.apache.avro.Schema.createRecord("nested" + levels, "doc ", "ns", false, fields);
     org.apache.avro.generic.GenericRecord lastLevelRecord = new GenericData.Record(lastLevelSchema);
@@ -251,8 +267,11 @@ public class Utils {
     lastLevelRecord.put(
         "level" + levels + "Array",
         List.of("level" + levels + "_" + "1", "level" + levels + "_" + "2"));
-    lastLevelRecord.put("level" + levels + "StringWithProps", "level" + levels + "_" + "WithProps");
+    lastLevelRecord.put(
+        "level" + levels + "StringWithPropsAndAlias", "level" + levels + "_" + "WithProps");
     lastLevelRecord.put("level" + levels + "Union", "level" + levels + "_" + "2");
+    lastLevelRecord.put("level" + levels + "Null", null);
+    lastLevelRecord.put("level" + levels + "NullRecord", null);
 
     org.apache.avro.Schema nextLevelSchema = lastLevelSchema;
     org.apache.avro.generic.GenericRecord nextLevelRecord = lastLevelRecord;
@@ -277,6 +296,8 @@ public class Utils {
       nextLevelRecord = currentLevelRecord;
     }
 
+    nextLevelSchema.addProp(AVRO_READ_OFFSET_PROP, 5);
+    nextLevelSchema.addProp("t1", 9);
     List<Field> pulsarFields =
         fields.stream().map(v -> new Field(v.name(), v.pos())).collect(Collectors.toList());
     return new GenericAvroRecord(new byte[0], nextLevelSchema, pulsarFields, nextLevelRecord);

